serve/mlc_serve/model/model_common.py

```diff
@@ -36,6 +38,59 @@ def get_num_cache_blocks(
     )
 
 
+def get_raw_logprob_info(
+    logits,
+    token,
+    top_logprobs_num,
+) -> RawLogprobsInfo:
+    logprobs = torch.log_softmax(logits, dim=-1)
+    res_logprob = logprobs[token].cpu().numpy()
+
+    if top_logprobs_num == 0:
+        top_logprobs = None
+        top_tokens = None
+    else:
+        assert top_logprobs_num <= LOGPROB_TOP_K_MAX, "Invalid input top_logprobs"
+        top_logprobs, top_tokens = torch.topk(
+            logprobs, k=top_logprobs_num, dim=-1, largest=True, sorted=True
+        )
+        top_tokens=top_tokens.cpu().numpy()
+        top_logprobs=top_logprobs.cpu().numpy()
+
+    # Set to raw logprob info
+    return RawLogprobsInfo(
+        # TODO(vvchernov): it is number, cpu().numpy()?
+        current_token=token.cpu().numpy(),
+        current_logprob=res_logprob,
+        top_tokens=top_tokens,
+        top_logprobs=top_logprobs,
+        previous_tokens=None
+    )
+
+
+def get_masked_logprobs(
+    logprob_infos: List[Optional[RawLogprobsInfo]],
+    mask: torch.Tensor,
+    sampling_params: List[SamplingParams],
+    logits: torch.Tensor,
+    tokens: torch.Tensor,
+) -> List[Optional[RawLogprobsInfo]]:
+    num_seq = len(logprob_infos)
+
+    mask_counter = 0
+    for i in range(num_seq):
+        if mask[i]:
```

Comment:
We don't want to loop over `num_seq` and check mask. I think `mask` is on GPU, so every time you access its element from python you get `cudaMemcpy`.

Please follow my suggestion https://github.com/octoml/mlc-llm/pull/82#discussion_r1467287183 instead. Rather than using `mask_greedy` or `mask_random`, collect a list of indices that wants logprob. Then you only need to loop over such indices here without mask check.

Reply:
> I think mask is on GPU

It is not so. Due to this I've prepared fix #178
I'm investigating the issue with performance degradation